portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -39,24 +42,33 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
     PreparedQuery results = datastore.prepare(query);
 
-    ArrayList<String> commentSection = new ArrayList<String>();
+    List<Map.Entry<String,String>> commentSection = new ArrayList<>();
     for (Entity entity : results.asIterable()) {
-        String text = (String) entity.getProperty("text");
-        commentSection.add(text);
+      String name = (String) entity.getProperty("name");
+      String text = (String) entity.getProperty("text");
+      Map.Entry<String,String> nameAndComment=new AbstractMap.SimpleEntry<>(name,text);
+      commentSection.add(nameAndComment);
     }
 
     int numberComments = getNumberComments(request);
     
-    ArrayList<String> limitedComments = new ArrayList<String>();
+    List<Map.Entry<String,String>> limitedComments = new ArrayList<>();
+
+    if (numberComments > commentSection.size()) {
+      numberComments = commentSection.size();
+    }
+
     for (int i = 0; i < numberComments; i++) {
-        limitedComments.add(commentSection.get(i));
+      String name = commentSection.get(i).getKey();
+      String text = commentSection.get(i).getValue();
+      Map.Entry<String,String> nameAndComment=new AbstractMap.SimpleEntry<>(name,text);
+      limitedComments.add(nameAndComment);
     }
```

Comment:
Made the comment on another PR of yours pointing out how you can avoid having to do this work by adding a SELECT TOP or LIMIT to your query, but in addition to that.. let's say that wasn't an option and you still received all of the comments in `results`. You could merge both of your for loops into just one, to avoid duplicating code and operations:

```
int readCommentCount = 0;
for (Entity entity : results.asIterable()) {
  if (readCommentCount >= numberComments) {
    break;
  }

  readCommentCount++;

  ... etc
}
```

It's not pretty but it's a bit better than doing the work twice.